### ISSUE TITLE
TaintingChecker false-positive for generics

### ISSUE THREAD

**Issue by AFulgens** [typetools/checker-framework#660]
Example:

```
public class Test {

  private final @Untainted Set<@Untainted String> s;

  public Test() {
    s = new @Untainted TreeSet<@Untainted String>();
  }
}
```

is correct (with Java 7 warning for redundant type specification), while

```
public class Test {

  private final @Untainted Set<@Untainted String> s;

  public Test() {
    s = new @Untainted TreeSet<>();
  }
}
```

emits `assignment.type.incompatible`, the TaintingChecker thinks that the `TreeSet` is `@Tainted`.

Note, that `new @Untainted TreeSet</*@Untainted String*/>();` also emits the false-positive, and it's also present if the generic type parameter does not have any annotations.

**Comment by AFulgens** [comment]
Related issue for `argument.type.incompatible`, for a function with a signature:
```
public @Untainted String doSomething(@Untainted Set<@Untainted String> input) {...}
```
the following invocation emits the warning:
```
@Untainted String s = doSomething(new @Untainted TreeSet<>());
```

**Comment by mernst** [comment]
The Tainting Checker issues this warning (for the code snippet in the first comment):
```
TaintingTest.java:9: error: [assignment.type.incompatible] incompatible types in assignment.
    s = new @Untainted TreeSet<>();
        ^
  found   : @TAINTED TreeSet<@UNTAINTED String>
  required: @UNTAINTED Set<@UNTAINTED String>
1 error
```

**Comment by mernst** [comment]
This appears to only be a problem for the Classic Tainting Checker, not the one built on the qualifier framework.
This problem does not have to do with caching.

I have created a test case in file
checker/tests/tainting_classic/TaintingDiamondInference.java (see commit c578bee72cf4196390b48517464c070b982a2807).

**Comment by wmdietl** [comment]
#2413 probably also fixed this, but the `TaintingDiamondInference.java` test produces some `cast.unsafe.constructor.invocation` warnings.
